zebra-network/src/address_book.rs

```diff
@@ -314,6 +326,34 @@ impl AddressBook {
         meta_addr
     }
 
+    /// Returns true if `updated` has a last_connection_state of `Responded`
+    /// and there is either:
+    /// - no previous addr with the updated.addr IP, or
+    /// - updated has a more recent last_response time than the existing [`MetaAddr`]
+    ///   for that IP in `most_recent_by_ip`.
+    fn should_update_most_recent_by_ip(&self, updated: MetaAddr) -> bool {
+        updated.last_connection_state == PeerAddrState::Responded
```

Comment:
> Are these invariants already documented, implemented, and tested somewhere else?

It's the `MetaAddr` that's most likely to fail the `!MetaAddr.has_connection_recently_responded()` check in `MetaAddr.is_ready_for_connection_attempt()`. I'll add that to the method's docs.

The redundant `self.last_connection_state == PeerAddrState::Responded` is from `was_recently_live()`.

Reply:
Hmm, I'm not sure we're talking about the same things here.

What I'm thinking of is documentation like:

> This code only needs to check a single cache entry, rather than the entire address book, because other code maintains these invariants:
> * the the only updates which change last_response are updates to the Responded state
> * the only field checked by was_recently_live() is last_response

Reply:
Added in https://github.com/ZcashFoundation/zebra/pull/7029/commits/0c769b657fbe1f66dd20b7ed874ce4b529395395.